File: presentation/src/main/java/android/montreal/devfest/presentation/navigation/NavigationModule.java
package android.montreal.devfest.presentation.navigation;

import android.content.Context;
import android.montreal.devfest.dagger.ActivityContext;

import javax.inject.Singleton;

import dagger.Binds;
import dagger.Module;
import dagger.Provides;

public interface NavigationModule {

    @Module
    interface ViewMore {
        @Binds
        Navigator.ViewMore viewMore(NavigatorWrapper wrapper);
    }

    @Module
    interface Website {
        @Binds
        Navigator.Website website(NavigatorWrapper wrapper);
    }

    @Module
    abstract class App {
        @Provides
        @Singleton
        static AppNavigator appNavigator() {
            return new AppNavigatorImpl();
        }
    }

    @Module
    abstract class Wrapper {
        @Provides
        static NavigatorWrapper wrapper(
                @ActivityContext Context context,
                AppNavigator appNavigator
        ) {
            return new NavigatorWrapper(context, appNavigator);
        }
    }
}

File: stuff_presentation/src/main/java/android/montreal/devfest/stuff/presentation/navigation/NavigationModule.kt
package android.montreal.devfest.stuff.presentation.navigation

import android.content.Context
import android.montreal.devfest.dagger.ActivityContext
import dagger.Binds
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

interface NavigationModule {

    @Module
    abstract class ViewMore {
        @Binds
        internal abstract fun viewMore(wrapper: NavigatorWrapper): Navigator.ViewMore
    }

    @Module
    abstract class Website {
        @Binds
        internal abstract fun website(wrapper: NavigatorWrapper): Navigator.Website
    }

    @Module
    object App {
        @Provides
        @Singleton
        @JvmStatic
        internal fun appNavigator(): AppNavigator {
            return AppNavigatorImpl()
        }
    }

    @Module
    object Wrapper {
        @Provides
        @JvmStatic
        internal fun wrapper(
                @ActivityContext context: Context,
                appNavigator: AppNavigator
        ): NavigatorWrapper {
            return NavigatorWrapper(context, appNavigator)
        }
    }
}
